### PR TITLE
Adapt to linear solver api from ewoms

### PR DESCRIPTION
**Pull request by totto82**
This PR
1) Adapt the ISTLSolverEbos class to the ewoms linear solver interface. i.e. adds methods prepareRhs(...), prepareMatrix(...) and solve(...)
2) Sets the linear solver to FlowIstlSolver and start calling it from the BlackoilModelEbos 
3) Moves makeOverlapRowsInvalid and WellModelMatrixAdapter from BlackoilModelEbos to ISTLSolverEbos

The PR should not give changes in results nor performance for any of the models.

### CODE REVIEW

**Comment by atgeirr** [comment]
> Moves makeOverlapRowsInvalid and WellModelMatrixAdapter from BlackoilModelEbos to ISTLSolverEbos

Why is the WellModelMatrixAdapter moved?

**Comment by atgeirr** [comment]
> PR #1617 moves the WellModelMatrixAdapter to a separate file opm/autodiff/MatrixBlock.hpp
> Maybe that is the best move?

I think perhaps so, but to answer your question:

> Where does the schur complement handling of well belong?

I think my answer is, one of two, either: 1. in the problem/model/whatever layer. Assuming a linear solver should not require knowledge about the domain or about the fact that we are actually coupling separate models (well + reservoir), it should just receive its operators/matrices and be happy. Or: 2. in the linear solver. Since there are preconditioners such as CPR that do require domain knowledge, that is already out of the bag. And if we want to solve group controls implicitly, we will need to extend the current system to account for that too (by explicit matrix terms or Schur complement again). Number 3 (well model) is right out, that model should just assemble the matrices, and the other layers should apply them as needed... I guess I am saying that we should change the well model too, from providing the abstract operation to providing the matrix needed to do it?

**Comment by andlaus** [comment]
note that option 2 seriously limits flexibility w.r.t. linear solvers because they need to be able to deal with non-matrix-only linear operators. I'm not sure if e.g., PETSc can do this and if it is efficient.

option 1 changes the results slightly and might be a bit slower. (I've been told. this can be tested using `--matrix-add-well-contributions=true`). Nevertheless I think that it is preferable. I agree that option 3 is the worst.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by hnil** [comment]
I think the option 3 is the best if I understand it correctly
  1) it make it easy to couple systems of genneral type, but and over the complexity of solving it to linear  system manipulation.
   2) It will be easy to choose to solve the full system or a reduced linear system. Even static nonlinear reduction would be easily done in this way (due to AD.)

Side question: What do they do in dune-fem for coupled equations like mixed formulations.

**Comment by hnil** [comment]
I think we should add this solveJacobian as an option like it is done by RK (there called solveJacobianEbos). Then this genneral method can be tested while the default option still live. I will also whould like to let the linear solver be gennerated as in a linear solver factory as before so we can choose run-time.

**Comment by andlaus** [comment]
@hnil: to be clear: this PR does *not* change the linear solver code itself, but merely unifies the API this code is called with the linear solver backends in eWoms. Essentially the API comes down to this:

```
LinearSolverBackend lsb(simulator);
// ...
lsb.prepare(A, b);
lsb.solve(x);
```

I don't think that this is very complicated. but I see that the PR is currently not mergeable. @totto82: who shall fix this? you or me?

**Comment by atgeirr** [comment]
Just some more thoughts here since I now understand it better (I think).

There is little or no value at this point in unifying the API with the ewoms linear solvers, due to the very essential difference in solve(): ISTLSolver will not deal with the matrix it is passed, but invoke the well model to do its operator. That is not the behaviour that can be expected/wanted from a general solver component. In the discussion above it looks like also @andlaus agreed that doing the Schur complement in the linear solver part is unwanted, so then we should not merge this, because it hides the uncommon behaviour (doing the Schur complement itself) behind the common interface.

I believe that the interface itself (`prepare()` then `solve()`) is fine, but doubt that it can be made general enough to support passing arbitrary operators (while ISTL supports this, and other backends might, it is not likely to be done in compatible ways) and keeping efficient. Therefore, we should make two separate code paths instead. One that just uses the ISTLSolver as it is now (or after this PR for that matter), and another that invokes a regular ewoms linear solver part, as in @dr-robertk's branch. We can let this be a simple runtime choice, letting it default to choosing the ISTLSolver. Then we can easily experiment with solver and preconditioner variants while keeping it all in one program.

Ultimately we should have a layer that deals with the coupling of equations and deals with the Schur complements. It would not be hard to design for ISTL, but being able to deal with arbitrary backends makes it a tougher problem.

**Comment by andlaus** [comment]
It sounds your comment is more an argument for always including the Schur component explicitly than against unifying the linear solver API, i.e., making `--matrix-add-well-contributions=true` hardcoded?!

**Comment by andlaus** [comment]
(which would anyway be a good idea if the preconditioner ought to consider the Schur complement.)

**Comment by atgeirr** [comment]
> It sounds your comment is more an argument for always including the Schur component explicitly than against unifying the linear solver API, i.e., making `--matrix-add-well-contributions=true` hardcoded?!

Then we would lose the ability to do it on the fly, which would be bad, or at least a major change.

**Comment by andlaus** [comment]
> which would be bad

why? I think it would be good because we would get rid of another obscure fiddling knob. (This statment assumes that the performance does not degrade too much for most decks.)

> or at least a major change.

Well, this is in the eye of the beholder, see #1702: 4 lines added, 80 lines removed.

**Comment by andlaus** [comment]
yes, but this might have had easy-to-fix reasons. I guess we should look into that.

**Comment by atgeirr** [comment]
> yes, but this might have had easy-to-fix reasons. I guess we should look into that.

But we do not have to do that as long as we simply keep the existing code operational. Why do we want the ability to use the ewoms linear solver interface in the first place? Only to get flexibility for experimentation. No need to kill the existing code for that, is my point.

**Comment by andlaus** [comment]
> But we do not have to do that as long as we simply keep the existing code operational.

which is exactly what this patch does.

> Why do we want the ability to use the ewoms linear solver interface in the first place? 

It would reduce code duplication considerably. judging from the past, this always has been a good idea. Also, do you want to redo the work for the linear solver backends which was already done for more research-y projects that are based on eWoms. Finally, flow does not have some interesting features at the moment; e.g., it does not implement algebraic overlaps which could seriously cripple AMG performance for parallel runs.

> No need to kill the existing code for that, is my point.

agreed. Exactly this is the purpose of the present PR.

**Comment by atgeirr** [comment]
> It would reduce code duplication considerably

This is not accomplished by this PR, and note that I *do* want the ewoms linear solver API available, as I wrote earlier. After this PR the ISTLSolver *looks like* it has the ewoms API, but it actually does not, since it does the well operator behind the scenes. I therefore do not see the point.

**Comment by andlaus** [comment]
> This is not accomplished by this PR

it is once you try to test-drive alternative linear solvers or  if you want to check out the behaviour with algebraic overlaps!

**Comment by andlaus** [comment]
> After this PR the ISTLSolver looks like it has the ewoms API, but it actually does not, since it does the well operator behind the scenes. I therefore do not see the point.

the point is that you switch to your custom linear solver, then run `flow` with `--matrix-add-well-contributions=true`. as far as I can see, this is still considerably easier than having to implement everything in opm-simulators!?

**Comment by totto82** [comment]
> Only to get flexibility for experimentation.

The whole point of this PR is to unify the linear interface within OPM. I.e to make it possible to run flow with linear solvers from ewoms and ebos with the ISTLSolver from opm-simulators. In the future, I hope they can live in peace in the same directory in the same repository.  Currently, the prepare method needs a matrixAdapter and creates the operator internally. Should we make the interface more general by instead passing the operator? Then we could do the well adaptation before passing it to the linear solver.

**Comment by atgeirr** [comment]
> Currently, the prepare method needs a matrixAdapter and creates the operator internally. Should we make the interface more general by instead passing the operator? Then we could do the well adaptation before passing it to the linear solver.

That sounds like a good idea, but can it be done for other backends than ISTL?

**Comment by atgeirr** [comment]
> The whole point of this PR is to unify the linear interface within OPM. I.e to make it possible to run flow with linear solvers from ewoms and ebos with the ISTLSolver from opm-simulators. In the future, I hope they can live in peace in the same directory in the same repository.

Again, the interface is not meaningfully unified, since in spite of using the prepare() and solve() method names, you cannot replace the ISTLSolver with another solver directly: you still need to know ahead of time where you are doing the Schur complement, which differs between the (changed) ISTLSolver and other solvers using the same interface. So there still must be an if-statement in solveJacobianSystem() or equivalent if we are to have the flexibility of doing both.

**Comment by andlaus** [comment]
> you cannot replace the ISTLSolver with another solver directly: you still need to know ahead of time where you are doing the Schur complement, which differs between the (changed) ISTLSolver and other solvers using the same interface. So there still must be an if-statement in solveJacobianSystem() or equivalent if we are to have the flexibility of doing both.

changing the linear solver is supposed to strictly be a developer only option. i.e., you are supposed to know what you're doing. what's wrong with making the live of these people easier?

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 1]
```diff
@@ -51,11 +52,112 @@ NEW_PROP_TAG(Scalar);
...
+public:
+  typedef M matrix_type;
+  typedef X domain_type;
+  typedef Y range_type;
+  typedef typename X::field_type field_type;
+
+#if HAVE_MPI
+  typedef Dune::OwnerOverlapCopyCommunication<int,int> communication_type;
+#else
+  typedef Dune::CollectiveCommunication< Grid > communication_type;
```
Comment (atgeirr) [inline opm/autodiff/ISTLSolverEbos.hpp]: Grid is not defined here.

File: opm/autodiff/ISTLSolverEbos.hpp [hunk 2]
```diff
@@ -381,10 +528,48 @@ namespace Opm
             }
         }
     protected:
+
+        bool isParallel() const {
+            return parallelInformation_.type() == typeid(ParallelISTLInformation);
```
Comment (atgeirr) [inline opm/autodiff/ISTLSolverEbos.hpp]: Needs #if HAVE_MPI for the ParallelISTLInformation class to exist, #else branch can return false.

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/ISTLSolverEbos.hpp (hunk 2 above)
Comment (totto82) [reply opm/autodiff/ISTLSolverEbos.hpp]: Thanks for the catch

File: opm/autodiff/ISTLSolverEbos.hpp (hunk 1 above)
Comment (totto82) [reply opm/autodiff/ISTLSolverEbos.hpp]: Fixed

**Comment by atgeirr** [comment]
jenkins build this serial please

**Comment by atgeirr** [comment]
All green, including serial. I'll merge this and then we can make the flexibility-adding step afterwards...

**Pull request merged**